### ISSUE TITLE
GUI shims run from the CLI don't run as child of the console

### ISSUE THREAD

**Issue by genotrance** [chocolatey/shimgen#42]
If you install a GUI app with Chocolatey, it generates a GUI shim as expected. Running this GUI shim executable from a cmd.exe or powershell.exe results in the shim running without a parent process.

Example:-
- Run putty.exe directly from CLI and it runs with cmd.exe as it's parent
- Run putty.exe shim from the CLI and it runs without a parent

![shim](https://user-images.githubusercontent.com/192231/36682381-8e29feaa-1ae0-11e8-90b0-5de750d89229.JPG)

For normal apps, this isn't an issue but for my proxy server which uses AttachConsole() to write to the console (if invoked from one), it [doesn't work](https://github.com/genotrance/px/issues/26) since it cannot find it's parent console.

Opening this issue since you'd expect the shim to run as a child process just like the application it is shimming.

**Comment by ferventcoder** [comment]
CLI shims always wait for exit, GUI shims always exit immediately. However if you need it to do something else, you could call `--shimgen-help` to learn of your options.

**Comment by ferventcoder** [comment]
As in `putty --shimgen-help`

**Comment by ferventcoder** [comment]
A full list of arguments you can call are at https://github.com/chocolatey/shimgen#shim-arguments as well

**Comment by genotrance** [comment]
Thank you for your input.

**Comment by ferventcoder** [comment]
We should probably leave this open as an enhancement until there is a way to create the shim in this way instead of a runtime call.

**Comment by genotrance** [comment]
The only sensible way to implement this would be to run a GUI app with wait for exit automatically set if it was run from a console. That way, even though both the app and the shim both run in the background, there's still a process hierarchy for apps like Px to look for a parent console and attach to it if desired.

**Comment by ferventcoder** [comment]
> The only sensible way to implement this would be to run a GUI app with wait for exit automatically set if it was run from a console.

@genotrance for your particular case that might be true. However for most folks they don't want to have a console that is blocked. If they are using it to open something like gitk or something they might have open for a long time, if it blocks on the console, then they are unable to use the console until close the GUI app. 

And for your particular use case, we should support a way of having the shim generated in a way that gives you the ability to specify.

**Comment by ferventcoder** [comment]
The defaults for shims have about 5+ years behind them and quite a bit of software. There are only a handful of edge cases that we've run into.

**Comment by genotrance** [comment]
I agree you need to be careful on how you proceed but I'm not saying that the shim should hold onto the console - that would make it into a console app. 

Assuming you make a GUI shim for a GUI app, what I'm asking for is the GUI shim to wait for the GUI app to exit if the GUI shim was invoked from a console. That way, the process hierarchy is maintained and the GUI app is free to attach to the parent console if required. The GUI shim would still run in the background and not hold onto the console, or block it.

I also agree that this is an edge case and you don't need to do it for every app but make it configurable.
